Feat(SPV-1546): delete user endpoint

dzolt-4chain opened the pull request:
<!-- 
    Thank you for contributing to our project! Before you submit your Pull Request, please make sure you've completed the items in this checklist.
    Please check the boxes below by putting an x in the [ ] like so: [x]. You can do it right after creating the PR.
-->

# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

<!-- 
## PR Title as Conventional Commit

Your PR title should also be a [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/). This will assist us in our release process and in assigning semantic version numbers to releases.
-->

<!-- 
## Related Tickets & Documents

If your changes relate to or close an issue, please include them below. We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For instance, having the text "closes #1234" would link the current pull request to issue number 1234. And when we merge the pull request, Github will automatically close the issue.

- Related Issue #
- Closes #
-->

<!--
## Feature/Issue Details

Your pull request should be tightly connected to a feature or issue. Ensure that you've covered all necessary pathways and there's a possibility to test code behavior following the issue description or feature specification.
-->

dzolt-4chain committed:
feat(SPV-1546): add delete user endpoint

dzolt-4chain committed:
test(SPV-1546): delete user endpoint tests

dzolt-4chain committed:
feat(SPV-1546): add api interface to service

dzolt-4chain committed:
test(SPV-1546): comment skip tests

dzolt-4chain committed:
test(SPV-1546): add test case to user delete

dzolt-4chain committed:
feat(SPV-1546): happy linter

dzolt-4chain committed:
Merge branch 'main' into feat/SPV-1546/delete_user_endpoint

chris-4chain commented on actions/v2/users/remove.go:
```diff
@@ -0,0 +1,27 @@
+		return
+	}
+
+	err = reqctx.Engine(c).UsersService().Remove(c.Request.Context(), userID)
```
You should rather use `s.usersService.Remove()`

dzolt-4chain replied:
ah right

chris-4chain commented on actions/v2/users/remove_test.go:
```diff
@@ -0,0 +1,174 @@
+			Post("/api/v2/admin/users")
+
+		// then:
+		then.Response(res).
+			HasStatus(201).
+			WithJSONMatching(`{
+				"id": "{{ matchAddress }}",
+				"createdAt": "{{ matchTimestamp }}",
+				"updatedAt": "{{ matchTimestamp }}",
+				"publicKey": "{{ .publicKey }}",
+				"paymails": [
+					{
+						"alias": "{{ .alias }}",
+						"avatar": "",
+						"domain": "example.com",
+						"id": "{{ matchNumber }}",
+						"paymail": "{{ .paymail }}",
+						"publicName": "{{ .publicName }}"
+					}
+				]
+			}`, map[string]any{
+				"publicKey":  publicKey,
+				"paymail":    userCandidate.DefaultPaymail(),
+				"publicName": userCandidate.DefaultPaymail().Alias(),
+				"alias":      userCandidate.DefaultPaymail().Alias(),
+			})
```
I would simplify this assertion to just `HasStatus(201)`. Because the creation of a user is already tested. The way I see it, this `remove_test.go` file should be primarily focused on `removing` and for related endpoints there should be detailed assertions.

dorzepowski replied:
agree

chris-4chain commented on actions/v2/users/remove_test.go:
```diff
@@ -0,0 +1,174 @@
+			SetPathParam("id", testState.userID).
+			Get("/api/v2/admin/users/{id}")
+
+		then.Response(res).
+			IsOK().
+			WithJSONMatching(`{
+				"id": "{{ matchAddress }}",
+				"createdAt": "{{ matchTimestamp }}",
+				"updatedAt": "{{ matchTimestamp }}",
+				"publicKey": "{{ .publicKey }}",
+				"paymails": [
+					{
+						"alias": "{{ .alias }}",
+						"avatar": "",
+						"domain": "example.com",
+						"id": "{{ matchNumber }}",
+						"paymail": "{{ .paymail }}",
+						"publicName": "{{ .publicName }}"
+					}
+				]
+			}`, map[string]any{
+				"publicKey":  publicKey,
+				"paymail":    userCandidate.DefaultPaymail(),
+				"publicName": userCandidate.DefaultPaymail().Alias(),
+				"alias":      userCandidate.DefaultPaymail().Alias(),
+			})
```
For getting user I would also minimize things to `IsOK()` assertion.

chris-4chain commented on actions/v2/users/server.go:
```diff
@@ -1,22 +1,37 @@
 	"github.com/rs/zerolog"
 )
 
+type usersService interface {
+	Create(ctx context.Context, newUser *usersmodels.NewUser) (*usersmodels.User, error)
+	Remove(ctx context.Context, userID string) error
+	Exists(ctx context.Context, userID string) (bool, error)
+	GetByID(ctx context.Context, userID string) (*usersmodels.User, error)
+	GetIDByPubKey(ctx context.Context, pubKey string) (string, error)
+	GetPubKey(ctx context.Context, userID string) (*primitives.PublicKey, error)
+	GetBalance(ctx context.Context, userID string) (bsv.Satoshis, error)
+}
```
You should add only methods that are actually used by this `APIUsers` struct. (Not every method that actual `UsersService` has).

I suppose `GetBalance` & `Remove` would be enough.

chris-4chain commented on actions/v2/users/server.go:
```diff
@@ -1,22 +1,37 @@
 	return APIUsers{
-		engine: engine,
-		logger: &logger,
+		usersService: engine.UsersService(),
```
Since you changed the way how the `APIUsers` is supposed to access "engine" methods. You should also adjust other methods like in `current.go`

chris-4chain commented on engine/v2/database/repository/users.go:
```diff
@@ -34,6 +34,35 @@ func (u *Users) Exists(ctx context.Context, userID string) (bool, error) {
 
+// Delete deletes user with userID and deletes their associated paymails, addresses, operations and tracked outputs
+func (u *Users) Delete(ctx context.Context, userID string) error {
+	error := u.db.WithContext(ctx).Unscoped().Transaction(func(tx *gorm.DB) error {
```
Variable 'error' collides with the 'builtin' interface.
Consider other name

dzolt-4chain committed:
feat(SPV-1546): address CR

dorzepowski committed:
test(SPV-1546): manual tests helper for delete user

dorzepowski commented on actions/v2/users/remove_test.go:
```diff
@@ -0,0 +1,133 @@
+			Post("/api/v2/admin/users")
+
+		// then:
+		then.Response(res).HasStatus(201)
```
### minor,
Can you add method if there isn't allready
`IsCreated` or `HasStatusCreated`?

dzolt-4chain replied:
added and changed in the other parts in repo

dorzepowski commented on actions/v2/users/remove_test.go:
```diff
@@ -0,0 +1,133 @@
+		testState.userID = getter.GetString("id")
+	})
+
+	t.Run("Get new user by id as admin", func(t *testing.T) {
```
we don't need this step

dzolt-4chain replied:
removed

dorzepowski commented on api/manualtests/userapi/delete_user_test.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/joomcode/errorx"
+)
+
+func TestDeleteCurrentUser(t *testing.T) {
```
@dzolt-4chain I added some manual test helper for deleting users

dzolt-4chain committed:
feat(SPV-1546): address cR

mrz1836 committed:
feat(errors): add client error handling framework

- Add errdef package for standardized error handling
- Implement client error builder and definitions
- Add HTTP error mapping and problem details
- Support error traits and properties

mrz1836 committed:
fix: minor fix for tests
